File: main/src/main/java/com/sedmelluq/discord/lavaplayer/natives/mp3/Mp3DecoderLibrary.java
package com.sedmelluq.discord.lavaplayer.natives.mp3;

import com.sedmelluq.discord.lavaplayer.natives.NativeLibLoader;

import java.nio.ByteBuffer;
import java.nio.ShortBuffer;

class Mp3DecoderLibrary {
  private Mp3DecoderLibrary() {

  }

  static Mp3DecoderLibrary getInstance() {
    NativeLibLoader.loadConnectorLibrary();
    return new Mp3DecoderLibrary();
  }

  native long create();

  native void destroy(long instance);

  native int decode(long instance, ByteBuffer directInput, int inputLength, ShortBuffer directOutput, int outputLengthInBytes);
}

File: main/src/main/java/com/sedmelluq/discord/lavaplayer/natives/mp3/Mp3DecoderLibrary.kt
package com.sedmelluq.discord.lavaplayer.natives.mp3

import com.sedmelluq.discord.lavaplayer.natives.ConnectorNativeLibLoader
import java.nio.ByteBuffer
import java.nio.ShortBuffer

internal class Mp3DecoderLibrary private constructor() {
    external fun create(): Long

    external fun destroy(instance: Long)

    external fun decode(
        instance: Long,
        directInput: ByteBuffer?,
        inputLength: Int,
        directOutput: ShortBuffer?,
        outputLengthInBytes: Int
    ): Int

    companion object {
        @kotlin.jvm.JvmStatic
        val instance: Mp3DecoderLibrary
            get() {
                ConnectorNativeLibLoader.loadConnectorLibrary()
                return Mp3DecoderLibrary()
            }
    }
}
